### DISCUSSION TITLE
Dynamic CPU Power Allocation Based on GPU Power

### DISCUSSION THREAD

**Discussion by XyKER1** [BartoszCichecki/LenovoLegionToolkit#1334] [Ideas]
I'm curious if this is possible because it would optimize performance greatly in CPU limited games. Basically what I'm asking is whether CPU power can be increased if the GPU drops below the TDP - **sort of like how the processor target power works**.

Now I know that I could set a higher CPU TDP, but I play Helldivers 2 where CPU usage increases as the game goes on, so it's good to have more GPU in the beginning but more CPU later on.

Let me give an example of this. I don't want my CPU + GPU power to exceed 230 W because then the battery will start draining. So right now, I have CPU PL1 and PL2 and 90 W and GPU at 140 W TDP. However, as the game goes on and CPU usage increases, the GPU will drop down to about 125 W, so ideally I would have the CPU then increase to 105 W to get a bit more FPS.
Now I know what you're thinking - why not just set the CPU TDP to 105 W? The problem with this is how inefficient Intel 13th gen CPUs are. They will consume 150+ W if they can during gaming even if it doesn't produce any extra FPS. So I only want the CPU power to go above 90 W if the GPU power drops below 140 W.
I also considered setting the GPU TDP to 125 W and then setting dynamic boost at 15 W so it could get to 140 W at the beginning. However, this isn't viable because the CPU power will run at 105 W if I set PL1/PL2 to that at all times, so the GPU would never hit 140 W. 

I know this sounds like a nitpicky thing, but I do think it could be very useful. I'm thinking there would probably have to be some sort of PL1/PL2 monitor that changes these every second based on GPU power. There's a baseline that the CPU can't go below (90 W in my case) and then the CPU power would increase for every watt of power less than the GPU TDP that the GPU is running at.

I completely understand if this isn't reasonable, but I thought there was no harm in throwing this out there just in case it's pretty easy to develop.

**Comment by BartoszCichecki** [comment]
I am not sure, if I understand it right, but the options that you see in Custom mode settings are the only ones that the BIOS/EC supports. They all are handled by laptops firmware. It isn't possible to change how they behave or add new ones.

**Reply by XyKER1** [reply]
Would it be possible though to check the GPU wattage, subtract it from the GPU TGP and then add this value to the CPU power limit every second using your own custom code? Many programs can read GPU wattage and/or change the CPU power limits, although obviously this might be quite difficult to do. I'm just saying that there are programs that can do each step in the process to add this feature, so it should be theoretically possible even if not explicitly supported by the EC.

As always, thank you for making this incredible program. There are so many user friendly features and I understand if this would be too much effort for something that seems so nitpicky. I'm just asking because this would be the last step needed to get complete control over  the CPU and GPU power behavior.

**Reply by BartoszCichecki** [reply]
No, as I said values that are configurable and configuration options of firmware.

**Reply by XyKER1** [reply]
Hi, I don't mean to sound rude or anything but setting CPU power limits in runtime is part of Legion's firmware (many programs such as this one, Throttlestop, Intel XTU, etc. can do it). In addition, reading the GPU's wattage is also part of the firmware and many programs such as Rivatuner, Nvidia GE, GPU-Z, etc. can do it.

Those are the only two values needed to do this power allocation - reading the GPU wattage and modifying the CPU's power limits accordingly.

**Reply by XyKER1** [reply]
Alternatively, is there a way to increase the number of custom mode profiles above 5. A little bit of a shortcut for me is to create a custom mode preset switch action that modifies the CPU and GPU power limits and I can toggle through these using F9. It's certainly not ideal, but it does work to some extent.

However, only having up to 5 custom mode presets seems like an unnecessary limit and I know they aren't stored in the firmware, so surely there is a way for you to increase this.

**Reply by BartoszCichecki** [reply]
Look, there are many ways to adjust power limits. LLT focuses on the interface that Lenovo provided and at the same time this is the only program except for Vantage that does it. The plan was always to expose Lenovo specific features, not to create a swiss army knife that can do everything.

Regarding number of custom profiles, yeah. 5 is just a random number I picked. I can remove it of course.

**Reply by BartoszCichecki** [reply]
Naah, I will just remove it - that is not a problem.

**Reply by XyKER1** [reply]
@BartoszCichecki Hey, just curious if there will be a release soon. It would be really helpful right now to have more than 5 profiles. I don't really know how to compile a program from the code - I tried running make.bat and clean.bat but nothing happened.

**Reply by Ace-Radom** [reply]
> @BartoszCichecki Hey, just curious if there will be a release soon. It would be really helpful right now to have more than 5 profiles. I don't really know how to compile a program from the code - I tried running make.bat and clean.bat but nothing happened.

@XyKER1 If you're just searching for a BETA version installer you can just go to the Lenovo Legion discord server (see README) and check `community contribution -> legion-toolkit` channal. You should see the BETA installers there.

If you really want to compile the source code, you need to install a couple of dev environments. Basically they're:

1. Visual Studio (someone told me that it is not really *necessary* for just compiling the src, but just for convenience u may install it)
2. .Net desktop development pack (In Visual Studio Installer)
3. .Net 8.0 runtime (LTS) (In Visual Studio Installer)
4. Inno Setup 6, if you want to create an installer on yourself

When you're done, run `make.bat`. If there's no error, you can just use the new compiled BETA version under `build` directory, or install with the installer under `build_installer`.
